spec/models/events/review_spec.rb

```diff
@@ -0,0 +1,118 @@
+# == Schema Information
+#
+# Table name: reviews
+#
+#  id              :bigint           not null, primary key
+#  body            :string
+#  status          :enum             default("active")
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#  github_id       :integer
+#  owner_id        :bigint
+#  pull_request_id :bigint           not null
+#
+# Indexes
+#
+#  index_reviews_on_owner_id         (owner_id)
+#  index_reviews_on_pull_request_id  (pull_request_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (owner_id => users.id)
+#  fk_rails_...  (pull_request_id => pull_requests.id)
+#
+
+require 'rails_helper'
+
+RSpec.describe Events::Review, type: :model do
+  describe 'validations' do
+    subject { build :review }
+
+    it 'is valid with valid attributes' do
+      expect(subject).to be_valid
+    end
+
+    it 'is not valid without github id' do
+      subject.github_id = nil
+      expect(subject).to_not be_valid
+    end
+
+    it { is_expected.to belong_to(:pull_request) }
+  end
+
+  describe 'actions' do
+    subject { create :review, payload: payload }
+
+    let(:payload) do
+      {
+        action: 'created',
+        review: {
+          pull_request_review_id: 237_895_671,
+          id: 284_312_630,
+          user: {
+            login: 'Codertocat',
+            id: 21_031_067,
+            node_id: 'MDQ6VXNlcjIxMDMxMDY3'
+          },
+          body: 'You might need to fix this.'
+        },
+        pull_request: {
+          id: 279_147_437
+        },
+        changes: {
+          body: 'Please fix this.'
+        }
+      }.deep_stringify_keys
+    end
+
+    describe '#find_or_create_review' do
+      subject { build :review }
+      before { create :pull_request, github_id: payload['pull_request']['id'] }
+
+      it 'creates a review' do
+        expect {
+          subject.send(:find_or_create_review, payload)
```

Comment:
This is a question. Would it be possible to call the method

`subject.find_or_create_review(payload)`

instead or is it a private method?

Reply:
Actually is private method, i understand that calling private methods in specs goes against TDD, but i didn't want to have to call the Event model to check that. If you have or know any other way to test that, it would be appreciated.

Reply:
How about something like

```
  before do
    def subject.find_or_create_review(payload)
      super(payload)
    end
  end
```

?

I just came up with it, I am not sure whether it could be convenient or not.

Reply:
Maybe an Rspec helper could make it explicit that the test intention is to send a private message to an object

```
before do
   enable_private_method :find_or_create_review, on: subject
end
```

Reply:
I will check if we are already handling this method in the specs without this one. Otherwise i guess we will should make a mock test that starts from the `Event` model where this method is triggered.

Reply:
@hdamico I took a deeper look at the context of this method call and I think there is change that might improve the separation of concerns and make the code a bit more clear.

If you look at the method

```
def find_or_create_review(payload)
    review_data = payload['review']
    review = Events::Review.find_or_create_by!(github_id: review_data['id']) do |rc|
    rc.owner = find_or_create_user(review_data['user'])
    rc.pull_request = Events::PullRequest.find_by!(github_id: payload['pull_request']['id'])
  end
  review.assign_attributes(payload: payload)
  review
end
```

you can see that even it is an instance method it does not access any instance side data from the receiver. That might point out that the method could be moved to a different place.

Also note that this method creates an instance of the same class than itself, which can be a bit confusing:

  `Review.new.new_from_payload` instead of `Review.new_from_payload`

So the method implementation suggests that it could be moved to the class side.

That would make it a bit clearer since classes responsibility is to create instances and would simplify the code.

In the Event class the method `Event.find_or_create_event` can be written as

```
def find_or_create_event
  const_event.new_from_payload(data)
end
```

and the method `Review.find_or_create_review` can be renamed and moved to the class side with

```
def self.new_from_payload(payload)
  comment_data = payload['comment']
  comment = self.find_or_create_by!(github_id: comment_data['id']) do |rc|
    rc.owner = User.new_from_payload(comment_data['user'])
    rc.pull_request = Events::PullRequest.find_by!(github_id: payload['pull_request']['id'])
  end
  comment.assign_attributes(payload: payload)
  comment
end
```

A side effect of this refactor is that it is no longer necessary to use `:send` and to dynamically build the method name in

```
def find_or_create_event
  const_event.new.send("find_or_create_#{name}", data)
end
```

since the message is replaced by the polymorphic method `new_from_payload` that each event type would implement.

There are several more small refactors that can be done to the method:

```
def self.new_from_payload(payload)
  # All these getters suggest that a service object to abstract the internal structure of
  # the payload might be a great improvement as you suggested in another comment.
  comment_data = payload['comment']
  comment_id = comment_data['id']
  user_data = comment_data['user']
  pull_request_id = payload['pull_request']['id']

  self.find_or_create_by!(github_id: comment_id) do |comment|
    comment.owner = User.new_from_payload(user_data)
    comment.pull_request = Events::PullRequest.find_by!(pull_request_id)
    comment.assign_attributes(payload: payload)
  end
end
```

but I think that for now the most significant change would be to move the creation of the Review instance to the class side.

Reply:
I know this kind of refactors put quite a delay on the merge of the original PR. Sorry about that but I think this refactor is significant.

Reply:
Great! That is a nice approach. Anyway I would change that in other pull request, actually i am making a refactor, and this will be useful.
Thanks!